keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -91,242 +97,358 @@ with the unreasonable development overhead of a webhook.
 
 We propose
 
-1. adding boolean vendor extensions to CRD OpenAPI schemas named `x-kubernetes-immutable` and `x-kubernetes-immutable-keys` with `true` as the only valid value.
-2. do **strict deep-equal** comparison of those fields marked as immutable during
-   update validation in the request pipeline, with these exceptions:
-   - for fields with `x-kubernetes-immutable-keys: true` only the keys of the map or array (specified via `x-kubernetes-list-map-keys`) are compared.
-   - for fields with `x-kubernetes-list-type: set` the order is ignored.
+1. adding two optional vendor extensions to CRD OpenAPI schemas:
+ 
+   1. `x-kubernetes-mutability: Immutable | AddOnly | RemoveOnly` and
+   2. `x-kubernetes-key-mutability: Immutable | AddOnly | RemoveOnly`.
+   
+   The second is restricted to `Immutable` *if* the schema is of type `object` and `properties` is set, the first is restricted otherwise.
```

Comment:
This says ```The second is restricted to `Immutable` *if* the schema is of type `object` and `properties` is set``` but https://github.com/sttts/enhancements/blob/sttts-immutable-fields-details/keps/sig-api-machinery/20190603-immutable-fields.md#immutableaddonlyremoveonly-object-keys-mutable-values says this is not allowed ```Because object properties in Kubernetes-like APIs should not be considered as a known set, but each field individually we disallow the use of x-kubernetes-key-mutability in this case. CRD validation will reject it.```

Reply:
That's the point: for objects with `properties` (= structs in Golang) we want to enforce that `x-kubernetes-mutability: Immutable | AddOnly | RemoveOnly` **on the fields** control whether a field can be added or removed. We don't want that through `x-kubernetes-key-mutability` on the parent struct.

Reply:
Let me reformulate this. It's too confusing.